pymc_marketing/clv/models/pareto_nbd.py

```diff
@@ -0,0 +1,589 @@
+        )
+        values = np.vstack((t_x.values, x.values)).T
+        loglike = pm.logp(pareto_dist, values).eval()
+        loglike = xarray.DataArray(data=loglike, dims=("chain", "draw", "customer_id"))
+
+        # TODO: Are these mean comparisons kosher?
+        if alpha.mean() < beta.mean():
+            min_of_alpha_beta, max_of_alpha_beta, p, _, _ = (
+                alpha,
+                beta,
+                r + x + n,
+                r + x,
+                r + x + 1,
+            )
+        else:
+            min_of_alpha_beta, max_of_alpha_beta, p, _, _ = (
+                beta,
+                alpha,
+                s + 1,
+                s + 1,
+                s,
+            )
+        abs_alpha_beta = max_of_alpha_beta - min_of_alpha_beta
+
+        log_p_zero = (
+            gammaln(r + x)
+            + r * log(alpha)
+            + s * log(beta)
+            - (gammaln(r) + (r + x) * log(alpha + T) + s * log(beta + T) + loglike)
+        )
+        log_B_one = (
+            gammaln(r + x + n)
+            + r * log(alpha)
+            + s * log(beta)
+            - (
+                gammaln(r)
+                + (r + x + n) * log(alpha + T + future_t)
+                + s * log(beta + T + future_t)
+            )
+        )
+        log_B_two = (
+            r * log(alpha)
+            + s * log(beta)
+            + gammaln(r + s + x)
+            + betaln(r + x + n, s + 1)
+            + log(
+                hyp2f1(
+                    r + s + x,
+                    p,
+                    r + s + x + n + 1,
+                    abs_alpha_beta / (max_of_alpha_beta + T),
+                )
+            )
+            - (gammaln(r) + gammaln(s) + (r + s + x) * log(max_of_alpha_beta + T))
+        )
+
+        def _log_B_three(i):
+            return (
+                r * log(alpha)
+                + s * log(beta)
+                + gammaln(r + s + x + i)
+                + betaln(r + x + n, s + 1)
+                + log(
+                    hyp2f1(
+                        r + s + x + i,
+                        p,
+                        r + s + x + n + 1,
+                        abs_alpha_beta / (max_of_alpha_beta + T + future_t),
+                    )
+                )
+                - (
+                    gammaln(r)
+                    + gammaln(s)
+                    + (r + s + x + i) * log(max_of_alpha_beta + T + future_t)
+                )
+            )
+
+        zeroth_term = (n == 0) * (1 - exp(log_p_zero))
+        first_term = n * log(future_t) - gammaln(n + 1) + log_B_one - loglike
+        second_term = log_B_two - loglike
+        third_term = logsumexp(
+            [
+                i * log(future_t) - gammaln(i + 1) + _log_B_three(i) - loglike
+                for i in range(n + 1)
+            ],
+            axis=0,
+        )
+
+        try:
+            size = len(x)
+            sign = np.ones(size)
+        except TypeError:
+            sign = 1
```

Comment:
When does this happen? Seems a bit dirty

Reply:
I'm guessing if someone runs a prediction on a single customer. I'll change this to an IF statement.